backend/dataall/modules/dataset_sharing/services/share_managers/s3_access_point_share_manager.py

```diff
@@ -140,34 +143,52 @@ def grant_target_role_access_policy(self):
         )
+        key_alias = f"alias/{self.dataset.KmsAlias}"
+        kms_client = KmsClient(self.dataset_account_id, self.source_environment.region)
+        kms_key_id = kms_client.get_key_id(key_alias)
+
         existing_policy = IAM.get_role_policy(
             self.target_account_id,
             self.target_requester_IAMRoleName,
-            "targetDatasetAccessControlPolicy",
+            IAM_ACCESS_POINT_ROLE_POLICY,
         )
         if existing_policy:  # type dict
-            if self.bucket_name not in ",".join(existing_policy["Statement"][0]["Resource"]):
-                logger.info(
-                    f'targetDatasetAccessControlPolicy exists for IAM role {self.target_requester_IAMRoleName}, '
-                    f'but S3 Access point {self.access_point_name} is not included, updating...'
-                )
-                target_resources = [
-                    f"arn:aws:s3:::{self.bucket_name}",
-                    f"arn:aws:s3:::{self.bucket_name}/*",
-                    f"arn:aws:s3:{self.dataset_region}:{self.dataset_account_id}:accesspoint/{self.access_point_name}",
-                    f"arn:aws:s3:{self.dataset_region}:{self.dataset_account_id}:accesspoint/{self.access_point_name}/*"
+            s3_target_resources = [
+                f"arn:aws:s3:::{self.bucket_name}",
+                f"arn:aws:s3:::{self.bucket_name}/*",
+                f"arn:aws:s3:{self.dataset_region}:{self.dataset_account_id}:accesspoint/{self.access_point_name}",
+                f"arn:aws:s3:{self.dataset_region}:{self.dataset_account_id}:accesspoint/{self.access_point_name}/*"
+            ]
+            share_manager = ShareManagerUtils(
+                self.session,
+                self.dataset,
+                self.share,
+                self.source_environment,
+                self.target_environment,
+                self.source_env_group,
+                self.env_group
+            )
+            share_manager.add_missing_resources_to_policy_statement(
+                self.bucket_name,
+                s3_target_resources,
+                existing_policy["Statement"][0],
+                IAM_ACCESS_POINT_ROLE_POLICY
+            )
+            if kms_key_id:
+                kms_target_resources = [
+                    f"arn:aws:kms:{self.dataset_region}:{self.dataset_account_id}:key/{kms_key_id}"
                 ]
-                existing_policy["Statement"][0]["Resource"].extend(target_resources)
-                policy = existing_policy
-            else:
-                logger.info(
-                    f'targetDatasetAccessControlPolicy exists for IAM role {self.target_requester_IAMRoleName} '
-                    f'and S3 Access point {self.access_point_name} is included, skipping...'
+                share_manager.add_missing_resources_to_policy_statement(
+                    kms_key_id,
+                    kms_target_resources,
+                    existing_policy["Statement"][1],
```

Comment:
IAM roles that were requesters and already had a `IAM_ACCESS_POINT_ROLE_POLICY`, that policy won't contain the second statement and will therefore fail. (This is actually the cause of error 3 in the testing). We need to add this KMS-statement before adding resources to it

Reply:
Good catch! Implemented solution in next revision.

Reply:
Code-wise the solution looks good. I will test but I'll mark this issue as resolved :)